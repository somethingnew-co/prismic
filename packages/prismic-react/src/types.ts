import React from 'react'
import { LinkResolver } from '@stnew/prismic-types'

interface Slice {
  (props: any): JSX.Element;
}

interface Slices {
  [key: string]: Promise<any> | Slice;
}

export interface Context {
  slices: Slices;
  linkResolver: LinkResolver;
  hrefResolver?: LinkResolver;
  rootResolver?: string;
}

export interface Provider extends Context {
  children: React.ReactNode;
}
