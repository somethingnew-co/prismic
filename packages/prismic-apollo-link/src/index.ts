export { default as prismicApolloLink } from './apolloLink'
