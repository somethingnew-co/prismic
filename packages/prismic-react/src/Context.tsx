import { createContext } from 'react'
import { Context } from './types'

const prismicContext: Context = {
  slices: {},
  linkResolver: () => '/',
  hrefResolver: () => '/',
  rootResolver: '',
}

const PrismicContext = createContext(prismicContext)

export default PrismicContext
