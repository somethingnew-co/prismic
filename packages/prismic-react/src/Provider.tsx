import React from 'react'
import PrismicContext from './Context'
import { Provider, Context } from './types'

function PrismicProvider({ slices, linkResolver, hrefResolver, children }: Provider): JSX.Element {

  const sliceMap = new Map(Object.entries(slices))

  const value: Context = {
    sliceMap,
    linkResolver,
    hrefResolver,
  }

  return <PrismicContext.Provider value={value}>{children}</PrismicContext.Provider>
}

export default PrismicProvider
