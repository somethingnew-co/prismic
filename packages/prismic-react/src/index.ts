export { default as PrismicContext } from './Context'
export { default as PrismicProvider } from './Provider'
export { default as usePrismic } from './usePrismic'
export { default as useHtmlSerializer } from './useHtmlSerializer'
export { SliceZone } from './SliceZone'
export { default as serializeElements } from './serializeElements'
export * from '@stnew/prismic'
