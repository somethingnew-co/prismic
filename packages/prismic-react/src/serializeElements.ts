import React from 'react'
import { Elements, HTMLSerializer } from 'prismic-reactjs'
import { HTMLSerializerElementMap } from './types'

function serializeElements(map: HTMLSerializerElementMap): HTMLSerializer<React.ReactNode> {
  return function (type, element, content, children, key): React.ReactNode {

    const node = map[type as Elements]

    if (node) {
      const [component, props] = node
      let propagator = props || {}

      if (typeof props === 'function') {
        propagator = props(element)
      }

      const propsWithKey = { ...propagator, key }

      // <img /> elements don't have children
      if (type === Elements.image) {
        return React.createElement(component, propsWithKey)
      }

      return React.createElement(component, propsWithKey, children)
    }
  }
}

export default serializeElements
