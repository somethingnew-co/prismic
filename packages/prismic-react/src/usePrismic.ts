import { useContext } from 'react'
import PrismicContext from './Context'
import { Context } from './types'

function usePrismic(): Context {
  const value = useContext(PrismicContext)
  return value
}

export default usePrismic
