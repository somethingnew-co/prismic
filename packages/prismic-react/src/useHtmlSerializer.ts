import React from 'react'
import { HTMLSerializerElementMap } from './types'
import { HTMLSerializer } from 'prismic-reactjs'
import serializeElements from './serializeElements'

function useHtmlSerializer(elementsMap: HTMLSerializerElementMap): HTMLSerializer<React.ReactNode> {
  return serializeElements(elementsMap)
}

export default useHtmlSerializer
