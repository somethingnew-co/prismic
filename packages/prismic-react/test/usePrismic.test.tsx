import { usePrismic } from '../src'

import React from 'react'
import { renderWithPrismicProvider } from './test-setup'
import { testDoc, slices as sliceData } from '../../../test/test-utils'

import { act } from '@testing-library/react'

function UsePrismicExample(): JSX.Element {
  const {
    sliceMap,
    linkResolver,
  } = usePrismic()

  const Slice = sliceMap.get('test_slice_1') as (props: any) => JSX.Element
  const data = sliceData[0]
  return (
    <>
      <Slice {...data} />
      <div>{linkResolver(testDoc)}</div>
    </>
  )
}

test('usePrismic', async () => {
  const promise = Promise.resolve()

  const { container } = renderWithPrismicProvider(<UsePrismicExample />)

  await act(() => promise)

  expect(container).toMatchSnapshot()
})
