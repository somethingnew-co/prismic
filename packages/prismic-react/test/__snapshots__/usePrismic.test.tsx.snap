// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePrismic 1`] = `
<div>
  <div>
    <h1>
      Hello World
    </h1>
  </div>
  <div>
    /test
  </div>
</div>
`;
